pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -445,26 +447,36 @@ func applyUpstreamTLSSettings(cluster *v2.Cluster, tls *networking.TLSSettings)
 	case networking.TLSSettings_MUTUAL, networking.TLSSettings_ISTIO_MUTUAL:
 		cluster.TlsContext = &auth.UpstreamTlsContext{
 			CommonTlsContext: &auth.CommonTlsContext{
-				TlsCertificates: []*auth.TlsCertificate{
-					{
-						CertificateChain: &core.DataSource{
-							Specifier: &core.DataSource_Filename{
-								Filename: tls.ClientCertificate,
-							},
-						},
-						PrivateKey: &core.DataSource{
-							Specifier: &core.DataSource_Filename{
-								Filename: tls.PrivateKey,
-							},
-						},
-					},
-				},
 				ValidationContextType: &auth.CommonTlsContext_ValidationContext{
 					ValidationContext: certValidationContext,
 				},
 			},
 			Sni: tls.Sni,
 		}
+		if meshConfig.SdsUdsPath == "" || tls.Mode == networking.TLSSettings_MUTUAL {
+			cluster.TlsContext.CommonTlsContext.TlsCertificates = []*auth.TlsCertificate{
+				{
+					CertificateChain: &core.DataSource{
+						Specifier: &core.DataSource_Filename{
+							Filename: tls.ClientCertificate,
+						},
+					},
+					PrivateKey: &core.DataSource{
+						Specifier: &core.DataSource_Filename{
+							Filename: tls.PrivateKey,
+						},
+					},
+				},
+			}
+		} else {
+			cluster.TlsContext.CommonTlsContext.TlsCertificateSdsSecretConfigs = []*auth.SdsSecretConfig{}
+			// tls.SubjectAltNames is set to upstreamServiceAccount for mTLS case.
+			for _, sa := range tls.SubjectAltNames {
```

Comment:
Is this correct to have one sds config for each SA?

Reply:
I think it's correct - 
1. here tls.SubjectAltNames is set to [upstreamServiceAccount](https://github.com/istio/istio/blob/3ff88a87cbfbd81ca789e2f83e9f2d40986578be/pilot/pkg/networking/core/v1alpha3/cluster.go#L233), which is got from [GetIstioServiceAccounts](https://github.com/istio/istio/blob/3ff88a87cbfbd81ca789e2f83e9f2d40986578be/pilot/pkg/networking/core/v1alpha3/cluster.go#L94); the array should contains only one value for one upstream servicehost/port. 
2. Also envoy support multiple sdsconfig(one service account maps to one secret)

Reply:
I don't really understand your reply. Do you mean tls.SubjectAltNames should have only one entry?

If not, then you have a list of sds configs. My quesitons are:
- Is this necessary, as all configs are almost identical (they differs by name only, which I don't think have any impact, see my other comments). 
- How client pick which config to use for a particular server, assuming we have multiple SAs. Also, do we really (want to) support it already?